libs/vkd3d/command.c

```diff
@@ -1839,6 +1839,120 @@ static void d3d12_command_list_invalidate_root_parameters(struct d3d12_command_l
     }
 }
 
+static void vk_access_and_stage_flags_from_d3d12_resource_state(const struct d3d12_device *device,
+        uint32_t state_mask, VkQueueFlags vk_queue_flags, VkPipelineStageFlags *stages, VkAccessFlags *access)
+{
+    VkPipelineStageFlags queue_shader_stages = 0;
+    uint32_t unhandled_state = 0;
+
+    if (vk_queue_flags & VK_QUEUE_GRAPHICS_BIT)
+    {
+        queue_shader_stages |= VK_PIPELINE_STAGE_VERTEX_SHADER_BIT |
+                VK_PIPELINE_STAGE_TESSELLATION_CONTROL_SHADER_BIT |
+                VK_PIPELINE_STAGE_TESSELLATION_EVALUATION_SHADER_BIT |
+                VK_PIPELINE_STAGE_GEOMETRY_SHADER_BIT |
+                VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT;
+    }
+
+    if (vk_queue_flags & VK_QUEUE_COMPUTE_BIT)
+        queue_shader_stages |= VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT;
+
+    if (state_mask == D3D12_RESOURCE_STATE_COMMON)
+    {
+        *stages |= VK_PIPELINE_STAGE_ALL_COMMANDS_BIT;
+        *access |= VK_ACCESS_MEMORY_READ_BIT | VK_ACCESS_MEMORY_WRITE_BIT;
+        return;
+    }
+
+    while (state_mask)
+    {
+        uint32_t state = state_mask & -state_mask;
+
+        switch (state)
+        {
+            case D3D12_RESOURCE_STATE_VERTEX_AND_CONSTANT_BUFFER:
+                *stages |= queue_shader_stages | VK_PIPELINE_STAGE_VERTEX_INPUT_BIT;
+                *access |= VK_ACCESS_UNIFORM_READ_BIT | VK_ACCESS_VERTEX_ATTRIBUTE_READ_BIT;
```

Comment:
One validation error I've seen in the past is that if we use this in the compute queue, we cannot use VERTEX_ATTRIBUTE_READ_BIT. Probably should check queue_flags for graphics related barriers.

Reply:
Should be fixed.